internal/jasmine_node_test/jasmine_runner.js

```diff
@@ -11,45 +12,113 @@ const UTF8 = {
+function watchMode(manifestLocation) {
+  // ibazel communicates with us via stdin
+  const rl = createInterface({input: process.stdin, terminal: false});
+  rl.on('line', chunk => {
+    if (chunk === IBAZEL_NOTIFY_BUILD_SUCCESS) {
+      // TODO: we should try keep the same instance of JasmineRunner if it's expensive to new up
+      // TODO: stop previous runs if we get multiple calls of this in quick succession (debounce)
+      const jrunner = createRunner(true);
+      addFiles(manifestLocation, jrunner);
+      jrunner.execute();
+    }
+  });
+  rl.on('close', () => {
+    // Give ibazel 5s to kill our process, otherwise do it ourselves
+    setTimeout(() => {
+      console.error('ibazel failed to stop jasmine after 5s; probably a bug');
+      process.exit(1);
+    }, 5000);
+  });
+}
+
+function deleteFromCache(fileName) {
+  // jasmine uses `require(fileName)` to read the files into the program
+  // so we have to delete them from the cache to ensure they read again on subsequent runs
+  // when in ibazel mode
+  delete require.cache[require.resolve(fileName)];
+}
+
+function addFiles(manifestLocation, jrunner){
+  deleteFromCache(manifestLocation)
+
+  const manifest = require.resolve(manifestLocation);
+  const file = fs.readFileSync(manifest, UTF8);
+  fs.readFileSync(manifest, UTF8).split('\n').filter(l => l.length > 0).forEach(f => {
+    // ensure it's actually read from disk again
+    deleteFromCache(f);
+    try {
+      // try requiring it incase there are syntax errors that kills the program later on
+      require(f);
+      jrunner.addSpecFile(f);
+    } catch (error) {
+      // dont add the files to jasmine since it'll run them and it'll kill the process
+      console.error('Error reading file: ', require.resolve(f))
+      console.error(error)
+    }
+  });
+}
 
+function createRunner(watch) {
   const jrunner = new JasmineRunner();
-  fs.readFileSync(manifest, UTF8)
-      .split('\n')
-      .filter(l => l.length > 0)
-      .forEach(f => jrunner.addSpecFile(f));
 
   var noSpecsFound = true;
   jrunner.addReporter({
     specDone: () => {
       noSpecsFound = false
     },
   });
+  
   // addReporter throws away the default console reporter
   // so we need to add it back
   jrunner.configureDefaultReporter({});
 
   jrunner.onComplete((passed) => {
     let exitCode = passed ? 0 : BAZEL_EXIT_TESTS_FAILED;
     if (noSpecsFound) exitCode = BAZEL_EXIT_NO_TESTS_FOUND;
-    process.exit(exitCode);
+    if (!watch) {
+      process.exit(exitCode)
+    }
   });
+  return jrunner;
+}
+
+function main(args) {
+  if (!args.length) {
+    throw new Error('Spec file manifest expected argument missing');
+  }
+  const manifestLocation = args[0];
+  // Remove the manifest, some tested code may process the argv.
+  process.argv.splice(2, 1)[0];
+
+  // only watch when under ibazel
+  const watch = Boolean(process.env[IBAZEL_NOTIFY_CHANGES]);
 
-  jrunner.execute();
-  return 0;
+  const jrunner = createRunner(watch);
```

Comment:
instead of passing a boolean here, how about passing an onComplete callback? that way anything to do with watch mode is localized here.
As you have it, there is stuff about exit codes in the onComplete above which is unused in single-run mode

Reply:
I've moved all of the watch logic into `iBazelMode`